src/index.ts

```diff
@@ -221,3 +221,18 @@ export async function isAvailable() {
 export function resetIsAvailableCache() {
   cachedIsAvailableResponse = undefined;
 }
+
+export function requestTimeout(): number {
+  // In testing, we were able to reproduce behavior similar to
+  // https://github.com/googleapis/google-auth-library-nodejs/issues/798
+  // by making many concurrent network requests. Requests do not actually fail,
+  // rather they take significantly longer to complete (and we hit our
+  // default 3000ms timeout).
+  //
+  // This logic detects a GCF environment, using the documented environment
+  // variables K_SERVICE and FUNCTION_NAME:
+  // https://cloud.google.com/functions/docs/env-var and, in a GCF environment
+  // extends timeouts signficantly (the 500,000ms value is based on testing
+  // 150 concurrent network connections.
+  return process.env.K_SERVICE || process.env.FUNCTION_NAME ? 500000 : 3000;
```

Comment:
max GCF timeout is 540000ms (default is 60000ms), at this point why not just bump it to the absolute max of 9 minutes? https://cloud.google.com/functions/docs/concepts/exec

Reply:
This timeout isn't actually tied to the execution time of functions, the underlying problem is that resources aren't available to a cloud function during initialization (they're available during execution). `500,000ms` is about double the time I was observing it take to execute `150` concurrent requests to the metadata server, my thinking was that this gives us a lot of head room.

Reply:
Is the thought that we do want to time out earlier if it's not running in GCF? It does seem like we'd want to wait for the metadata connection(s) even if they're not running in GCF, and that'd simplify things.

Reply:
@feywind in many environments, e.g., AWS, your personal computer, the metadata server might legitimately be not available. I mainly want to be cautious in case in some environments this manifests itself as a tar pit (requests taking a long time to time out).

I think 3s is a pretty safe assumption for a response, unless we're 99% sure we're in a serverless environment.